smdebug/tensorflow/keras.py

```diff
@@ -317,12 +318,28 @@ def _prepare_layers(self, mode):
 
     def _prepare_non_layer_tensors(self):
         # for gradients, optimizer_variables
+        custom_collections = set()
+        default_tf_collection = set()
+
         for coll in self.collection_manager.get_collections().values():
-            for tensor_ref in coll.get_tensors():
+            if coll.name not in DEFAULT_TF_COLLECTIONS:
+                custom_collections.add(coll)
+            else:
+                default_tf_collection.add(coll)
+
+        for default_coll in default_tf_collection:
+            for tensor_ref in default_coll.get_tensors():
                 if tensor_ref.name not in self.tensor_to_collections:
-                    self.tensor_to_collections[tensor_ref.name] = {coll}
-                elif coll not in self.tensor_to_collections[tensor_ref.name]:
-                    self.tensor_to_collections[tensor_ref.name].add(coll)
+                    self.tensor_to_collections[tensor_ref.name] = {default_coll}
+                elif default_coll not in self.tensor_to_collections[tensor_ref.name]:
+                    self.tensor_to_collections[tensor_ref.name].add(default_coll)
+
+                # Add tensor to custom collections
+                for custom_coll in custom_collections:
```

Comment:
is this really reqd ? 

So what happens when, say you add only custom collection("optimizer_Variable").include("Adam.*") and this change is not there ? 

Don't we see Adam* tensors getting saved?

Tensors should be saved only once even if they are in different collection. Collection A can have tensorA, tensorB 
and Collection B can have tensorA and tensorB , 
but total number of tensors that should be saved is tensorA, tensorB .

Reply:
The total number of tensors do not change with my changes.

Without my changes, the test fails for this particular case:

```
@pytest.mark.slow
def test_include_only_custom_collection(out_dir, tf_eager_mode):
    include_collections = ["custom_optimizer_variables"]
    save_config = SaveConfig(save_interval=3)
    hook = smd.KerasHook(
        out_dir,
        save_config=save_config,
        include_collections=include_collections,
        reduction_config=ReductionConfig(norms=ALLOWED_NORMS, reductions=ALLOWED_REDUCTIONS),
    )
    hook.get_collection("custom_optimizer_variables").include("Adam")
    helper_keras_fit(out_dir, hook=hook, steps=["train", "eval", "predict"], eager=tf_eager_mode)

    trial = smd.create_trial(path=out_dir)
    assert len(trial.tensor_names()) == (8 if is_tf_2_2() and tf_eager_mode else 9)
    assert len(trial.tensor_names(collection="custom_optimizer_variables")) == 5
```

Reply:
I have added this test to the test suite as well